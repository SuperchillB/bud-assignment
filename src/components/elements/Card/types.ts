import { DisplayProps, FlexboxProps, LayoutProps } from 'styled-system';
import BoxProps from '../Box/types';

export type CardProps = {
  children?: React.ReactNode;
} & FlexboxProps &
  LayoutProps;

export type CardContainerProps = BoxProps;

export type CardDisplayProps = {
  children: React.ReactNode;
} & DisplayProps &
  FlexboxProps;
