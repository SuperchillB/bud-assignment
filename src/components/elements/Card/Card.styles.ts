import styled from 'styled-components';
import { Flex } from '../Flex';
import { CardContainerProps } from './types';

export const CardContainer = styled(Flex)<CardContainerProps>`
  flex-direction: column;
  background-color: ${({ theme }) => theme.colors.white};
  border-radius: ${({ theme }) => theme.radii.card};
  box-shadow: ${({ theme }) => theme.shadows.card};
`;
