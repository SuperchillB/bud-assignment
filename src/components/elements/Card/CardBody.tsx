import React from 'react';
import { Box } from '../Box';
import { CardDisplayProps } from './types';

const CardBody = ({ children, ...props }: CardDisplayProps) => (
  <Box mb="xlarge" {...props}>
    {children}
  </Box>
);

export default CardBody;
