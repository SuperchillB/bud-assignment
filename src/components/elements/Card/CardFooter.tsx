import React from 'react';
import { Box } from '../Box';
import { CardDisplayProps } from './types';

const CardFooter = ({ children, ...props }: CardDisplayProps) => (
  <Box mt="auto" {...props}>
    {children}
  </Box>
);

export default CardFooter;
