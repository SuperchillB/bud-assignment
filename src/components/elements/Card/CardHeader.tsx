import React from 'react';
import { Box } from '../Box';
import { CardDisplayProps } from './types';

const CardHeader = ({ children, ...props }: CardDisplayProps) => (
  <Box mb="xlarge" {...props}>
    {children}
  </Box>
);

export default CardHeader;
