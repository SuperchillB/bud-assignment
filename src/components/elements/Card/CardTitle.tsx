import React from 'react';
import { Heading } from '../Heading';
import { CardDisplayProps } from './types';

const CardTitle = ({ children, ...props }: CardDisplayProps) => (
  <Heading mb="tiny" variant="h4" {...props}>
    {children}
  </Heading>
);

export default CardTitle;
