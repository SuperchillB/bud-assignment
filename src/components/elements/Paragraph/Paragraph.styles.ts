import styled from 'styled-components';
import {
  border,
  color,
  compose,
  flexbox,
  layout,
  position,
  space,
  typography,
  variant,
} from 'styled-system';
import { ParagraphProps } from './types';

const paragraphStyles = {
  mt: '0rem',
  mb: '0rem',
  fontWeight: 'normal',
};

export const Paragraph = styled.p<ParagraphProps>(
  variant({
    variants: {
      body: {
        ...paragraphStyles,
        fontSize: 'small',
      },
      subTitle: {
        ...paragraphStyles,
        fontSize: 'tiny',
      },
    },
  }),
  compose(space, color, layout, flexbox, border, position, typography),
);

Paragraph.defaultProps = {
  variant: 'body',
};
