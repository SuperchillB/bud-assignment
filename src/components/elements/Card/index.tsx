import theme from '../../../styles/theme';
import { CardContainer } from './Card.styles';
import CardBody from './CardBody';
import CardFooter from './CardFooter';
import CardHeader from './CardHeader';
import CardTitle from './CardTitle';
import { CardProps } from './types';

/*

Comments:
  - Assuming here Card component is not clickable and doesn't have a "clickable" prop
  
*/

const Card = ({ children, ...props }: CardProps) => {
  return (
    <CardContainer
      p={{ _: theme.space.large, md: theme.space.xlarge }}
      justifyContent="stretch"
      {...props}
    >
      {children}
    </CardContainer>
  );
};

export default Object.assign(Card, {
  Header: CardHeader,
  Title: CardTitle,
  Body: CardBody,
  Footer: CardFooter,
});
