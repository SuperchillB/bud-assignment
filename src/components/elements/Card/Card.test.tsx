import { screen as screenRTL } from '@testing-library/react';
import Card from '.';
import { renderWithClient } from '../../../tests/utils';
import { Paragraph } from '../Paragraph';

describe('Card component', () => {
  test('renders its children properly', () => {
    renderWithClient(
      <Card>
        <Card.Header>Test header</Card.Header>
        <Card.Body>
          <Card.Title>Test title</Card.Title>
          <Paragraph variant="body">Test body</Paragraph>
        </Card.Body>
        <Card.Footer>Test footer</Card.Footer>
      </Card>,
    );

    expect(screenRTL.getByText('Test header')).toBeInTheDocument();
    expect(screenRTL.getByText('Test body')).toBeInTheDocument();
    expect(screenRTL.getByText('Test footer')).toBeInTheDocument();
  });

  test('renders with given props', () => {
    const { container } = renderWithClient(
      <Card
        flex={[
          '0 0 45%',
          '0 0 45%',
          `0 1 calc(50% - 1rem)`,
          '1 calc(100% / 5)',
        ]}
        width={['auto', 'auto', '50%', 'auto']}
        maxWidth={['30rem', '30rem', 'none', '23.5rem']}
      >
        <Card.Header>Test header</Card.Header>
        <Card.Body>
          <Card.Title>Test title</Card.Title>
          <Paragraph variant="body">Test body</Paragraph>
        </Card.Body>
        <Card.Footer display={['block', 'block', 'none']}>
          Test footer
        </Card.Footer>
      </Card>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
