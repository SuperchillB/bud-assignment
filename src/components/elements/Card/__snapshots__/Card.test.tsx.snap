// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component renders with given props 1`] = `
<div
  class="sc-bczRLJ sc-gsnTZi sc-dkzDqf lmcTtE jDemKD fvrKEF"
  width="auto,auto,50%,auto"
>
  <div
    class="sc-bczRLJ hSnyMT"
  >
    Test header
  </div>
  <div
    class="sc-bczRLJ hSnyMT"
  >
    <h4
      class="sc-hKMtZM fVHIEU"
    >
      Test title
    </h4>
    <p
      class="sc-eCYdqJ kKjzFp"
    >
      Test body
    </p>
  </div>
  <div
    class="sc-bczRLJ eDNHpP"
    display="block,block,none"
  >
    Test footer
  </div>
</div>
`;
